### PR TITLE
e3c256d4i support

### PR DESCRIPTION
**Pull request by ofaurax**
By default, update is broken on e3c256d4i. This PR fixes it. Compatibility with e3c246d4i needs to be tested

### CODE REVIEW

File: providers/asrockrack/firmware.go [hunk 1]
```diff
@@ -74,7 +74,16 @@ func (a *ASRockRack) firmwareInstallBMC(ctx context.Context, reader io.Reader, f
 	a.log.V(2).WithValues("step", "2/4").Info("upload BMC firmware image")
 	err = a.uploadFirmware(ctx, "api/maintenance/firmware", reader, fileSize)
 	if err != nil {
-		return errors.Wrap(err, "failed in step 2/4 - upload BMC firmware image")
+		// Try the special URL for e3c256d4i
```
Comment (joelrebel) [inline providers/asrockrack/firmware.go:77]: Can we check what the error is - if it contains 404, then we attempt the other endpoint.

File: providers/asrockrack/firmware.go [hunk 2]
```diff
@@ -74,7 +74,16 @@ func (a *ASRockRack) firmwareInstallBMC(ctx context.Context, reader io.Reader, f
 	a.log.V(2).WithValues("step", "2/4").Info("upload BMC firmware image")
 	err = a.uploadFirmware(ctx, "api/maintenance/firmware", reader, fileSize)
```
Comment (joelrebel) [inline providers/asrockrack/firmware.go:75]: Would prefer we identify the device model number and call the appropriate endpoint,

The model number can be obtained through this method
https://github.com/bmc-toolbox/bmclib/blob/153e11d62a71723d41a908ad362ab5e6cfc4cd54/providers/asrockrack/inventory.go#L77

File: providers/asrockrack/helpers.go [hunk 3]
```diff
@@ -181,7 +181,15 @@ func (a *ASRockRack) createUpdateUser(ctx context.Context, account *UserAccount)
 // at this point all logged in sessions are terminated
 // and no logins are permitted
 func (a *ASRockRack) setFlashMode(ctx context.Context) error {
-	_, statusCode, err := a.queryHTTPS(ctx, "api/maintenance/flash", "PUT", nil, nil, 0)
+	// preserve config is needed by e3c256d4i
+	pConfig := &preserveConfig{PreserveConfig: 1}
```
Comment (joelrebel) [inline providers/asrockrack/helpers.go:185]: set this configuration based on device model number

**Review by joelrebel — CHANGES_REQUESTED** [review]
Thanks for this submission, can we also make sure this does not cause a regression on the older `e3c246d4i`

File: providers/asrockrack/helpers.go (hunk 3 above)
Comment (ofaurax) [reply providers/asrockrack/helpers.go]: OK, but this adds common as dependency

File: providers/asrockrack/firmware.go (hunk 2 above)
Comment (ofaurax) [reply providers/asrockrack/firmware.go]: Done

File: providers/asrockrack/firmware.go (hunk 1 above)
Comment (ofaurax) [reply providers/asrockrack/firmware.go]: We use the model now

File: providers/asrockrack/firmware.go [hunk 4]
```diff
@@ -63,16 +63,31 @@ func (a *ASRockRack) FirmwareInstallStatus(ctx context.Context, installVersion,
 func (a *ASRockRack) firmwareInstallBMC(ctx context.Context, reader io.Reader, fileSize int64) error {
 	var err error
 
+	// 0. take the model so that we use a different endpoint on E3C256D4ID-NL
+	device := common.NewDevice()
+	device.Metadata = map[string]string{}
+	err = a.fruAttributes(ctx, &device)
+	if err != nil {
+		a.log.V(2).WithValues("step", "0/4").Info("failed to get model, err: " + err.Error())
+		a.log.V(2).WithValues("step", "0/4").Info("continue with default endpoint")
```
Comment (joelrebel) [inline providers/asrockrack/firmware.go:72]: I suggest this returns an error here - if the model cannot be identified, its unlikely to be supported

File: providers/asrockrack/firmware.go [hunk 5]
```diff
@@ -63,16 +63,31 @@ func (a *ASRockRack) FirmwareInstallStatus(ctx context.Context, installVersion,
...
 	if err != nil {
 		return errors.Wrap(err, "failed in step 1/4 - set device to flash mode")
 	}
 
 	// 2. upload firmware image file
-	a.log.V(2).WithValues("step", "2/4").Info("upload BMC firmware image")
-	err = a.uploadFirmware(ctx, "api/maintenance/firmware", reader, fileSize)
+	fw_endpoint := "api/maintenance/firmware"
+	// E3C256D4ID-NL calls a different endpoint for firmware upload
+	if device.Model == "E3C256D4ID-NL" {
```
Comment (joelrebel) [inline providers/asrockrack/firmware.go:86]: Use `strings.EqualFold(device.Model, "E3C256D4ID-NL")`

File: providers/asrockrack/helpers.go [hunk 6]
```diff
@@ -181,7 +182,23 @@ func (a *ASRockRack) createUpdateUser(ctx context.Context, account *UserAccount)
...
 // and no logins are permitted
 func (a *ASRockRack) setFlashMode(ctx context.Context) error {
-	_, statusCode, err := a.queryHTTPS(ctx, "api/maintenance/flash", "PUT", nil, nil, 0)
+	device := common.NewDevice()
+	device.Metadata = map[string]string{}
+	_ = a.fruAttributes(ctx, &device)
+
+	pConfig := &preserveConfig{}
+	// preserve config is needed by e3c256d4i
+	if device.Model == "E3C256D4ID-NL" {
```
Comment (joelrebel) [inline providers/asrockrack/helpers.go:191]: use `strings.EqualFold()` here

**Review by joelrebel — CHANGES_REQUESTED** [review]

File: providers/asrockrack/firmware.go (hunk 4 above)
Comment (ofaurax) [reply providers/asrockrack/firmware.go:72]: Until now, we used the default endpoint for every model, and now only E3C256D4I needs a special one.
So, if the model is unknown, we still try the default one.

However, the code you mention test the error of `fruAttributes` and it's true that if we fail on getting them, we might have a real problem.

**Review by joelrebel — APPROVED** [review]
Thanks for the fixes!

**Pull request merged**